lib/elixir/lib/calendar/datetime.ex

```diff
@@ -889,6 +1075,46 @@ defmodule DateTime do
     Calendar.ISO.add_day_fraction_to_iso_days(iso_days, -offset, 86400)
   end
 
+  @no_valid_time_zone_database_error "No valid TimeZoneDatabase provided or configured. Configure with :elixir_config.put(:time_zone_module, module_name)"
+  @spec by_wall(TimeZoneDatabase.t(), Calendar.time_zone(), TimeZoneDatabase.gregorian_seconds()) ::
+          {:single, TimeZoneDatabase.time_zone_period()}
+          | {:ambiguous, [TimeZoneDatabase.time_zone_period()]}
+          | {:gap, [TimeZoneDatabase.time_zone_period()]}
+          | {:error, :time_zone_not_found}
+  defp by_wall(time_zone_data_module, time_zone, gregorian_seconds) do
+    time_zone_data_module = time_zone_data_module_from_parameter(time_zone_data_module)
+
+    try do
+      time_zone_data_module.by_wall(time_zone, gregorian_seconds)
+    rescue
+      UndefinedFunctionError ->
```

Comment:
I might move these checks to the function clause `time_zone_data_module_from_parameter(:from_config)`:

```elixir
defp time_zone_data_module_from_parameter(:from_config) do
  case :elixir_config.safe_get(:time_zone_module, :from_config) do
    time_zone_data_module -> time_zone_module
    nil -> raise "<A good error message>"
  end
end
```

If a `time_zone_data_module` is provided, it should implement the `TimeZoneDatabase` behaviour. In those cases, I think the error is quite easy to understand: `** (UndefinedFunctionError) function MyTimeZoneDatabase. by_utc/2 is undefined or private`.

Beside, I would use the same name for the variable (now `time_zone_data_module`) and for the key in the configuration (now `time_zone_module`).